addons/mail/tests/test_ir_ui_menu.py

```diff
@@ -0,0 +1,136 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from unittest.mock import Mock, patch
+
+from odoo.addons.mail.tests.common import MailCommon, mail_new_test_user
+from odoo.tests import users
+from odoo.tests.common import warmup
+
+
+class TestMenuRootLookupByModel(MailCommon):
+    """ Test the determination of the best menu root for a given model.
+
+    When sharing a record through a link, it doesn't contain the menu context
+    (menu root). To help the user, we try to restore a root menu related to
+    the record when redirecting to the record from the link. That's what is tested
+    here. For more details see IrUiMenu._get_best_backend_root_menu_id_for_model.
+    """
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        Menu = cls.env['ir.ui.menu']
+        Action = cls.env['ir.actions.act_window']
+
+        def _new_menu(name, parent_id=None, action=None):
+            inst = Menu.create({'name': name, 'parent_id': parent_id})
+            if action:
+                inst.action = action
+            return inst
+
+        new_menu = Mock(side_effect=_new_menu)
+
+        def new_action(name, res_model, view_mode=None, domain=None, context=None):
+            return Action.create({
+                'name': name,
+                'res_model': res_model,
+                'view_mode': view_mode,
+                'domain': domain,
+                'context': context or {},
+                'type': 'ir.actions.act_window',
+            })
+
+        # Remove all menus and setup test menu with known results
+        Menu.search([]).unlink()
+
+        menu_root_contact = new_menu('Contacts')
```

Comment:
Could be great to make a small table-docstring with the test menus and actions hierarchy so that it eases understanding of tests :)